sympy/tensor/array/expressions/array_expressions.py

```diff
@@ -1,34 +1,46 @@
+import itertools
 import operator
-from collections import defaultdict, Counter
+import typing
+from collections import Counter, defaultdict
 from functools import reduce
-import itertools
 from itertools import accumulate
-from typing import Optional, List, Dict as tDict, Tuple as tTuple
-
-import typing
+from typing import Dict as tDict
+from typing import Iterable, List, Optional
+from typing import Tuple as tTuple
+from typing import overload
 
+from sympy.combinatorics import Permutation
+from sympy.combinatorics.permutations import _af_invert
 from sympy.core.basic import Basic
 from sympy.core.containers import Tuple
 from sympy.core.expr import Expr
-from sympy.core.function import (Function, Lambda)
+from sympy.core.function import Function, Lambda
 from sympy.core.mul import Mul
+from sympy.core.numbers import One, Zero
 from sympy.core.singleton import S
 from sympy.core.sorting import default_sort_key
-from sympy.core.symbol import (Dummy, Symbol)
+from sympy.core.symbol import Dummy, Symbol
+from sympy.core.sympify import _sympify
 from sympy.matrices.expressions.diagonal import diagonalize_vector
-from sympy.matrices.expressions.matexpr import MatrixExpr
+from sympy.matrices.expressions.matexpr import MatrixElement, MatrixExpr
 from sympy.matrices.expressions.special import ZeroMatrix
-from sympy.tensor.array.arrayop import (permutedims, tensorcontraction, tensordiagonal, tensorproduct)
+from sympy.tensor.array.arrayop import (
+    permutedims,
+    tensorcontraction,
+    tensordiagonal,
+    tensorproduct,
+)
 from sympy.tensor.array.dense_ndim_array import ImmutableDenseNDimArray
+from sympy.tensor.array.expressions.utils import (
+    _apply_recursively_over_nested_lists,
+    _build_push_indices_down_func_transformation,
+    _build_push_indices_up_func_transformation,
+    _get_contraction_links,
+    _get_mapping_from_subranks,
+    _sort_contraction_indices,
+)
 from sympy.tensor.array.ndim_array import NDimArray
-from sympy.tensor.indexed import (Indexed, IndexedBase)
-from sympy.matrices.expressions.matexpr import MatrixElement
-from sympy.tensor.array.expressions.utils import _apply_recursively_over_nested_lists, _sort_contraction_indices, \
-    _get_mapping_from_subranks, _build_push_indices_up_func_transformation, _get_contraction_links, \
-    _build_push_indices_down_func_transformation
-from sympy.combinatorics import Permutation
-from sympy.combinatorics.permutations import _af_invert
-from sympy.core.sympify import _sympify
```

Comment:
is there a reason for changing the import lines so much?

Reply:
[isort](https://pycqa.github.io/isort). But ok, I'll kick that commit out

Reply:
Done :heavy_check_mark: 
But tbh, those unsorted, unorganised import statements in sympy give me goosebumps... :sweat_smile:

Reply:
Well, I am actually OK with reorganizing the imports. I usually don't edit those lines manually, as I use PyCharm's auto-import capabilities.

Maybe this discussion could apply to all of SymPy (and not just to this module)?

Reply:
(maybe reorganizing the imports can be done in a separate PR...)

Reply:
I looked this up once in the issues and vaguely remembered someone already tried applying isort (can't find that issue/PR anymore). Problem is that the order of imports matters in sympy: ordering the imports can cause circular imports (huge code smell imho :grimacing:).

At any rate, I don't think it's good to start applying isort on arbitrary modules without an official policy on this. That's why I kicked out that commit.

Reply:
There it is: https://github.com/sympy/sympy/pull/13007#issuecomment-316875190

Reply:
I think sorting imports could be a great idea, but should be done as a seperate issue/PR, especially considering work has been done on the imports recently (https://github.com/sympy/sympy/pull/22342).

Reply:
Ah yeah good point.

> I use PyCharm's auto-import capabilities.

As a side note, not a PyCharm user, but I'm quite sure it can call isort automatically upon saving as well.